### ISSUE TITLE
ctrl modifier not applied in certain situations (vmware horizon client window with Windows guest)

### ISSUE THREAD

**Issue by aerusso** [Xpra-org/xpra#4157]
**Describe the bug**
Pressing Ctrl+(key) only sends (key) when pressed inside the vmware horizon viewer client window (i.e., the window that actually shows the remote vm).  Yes, I am using xpra to view another remote-desktop viewer that is in-turn viewing a Windows guest OS.

**To Reproduce**
1. `xpra start :1 --start-child=vmware-view --exit-with-children --daemon=no --systemd-run=no -d keyboard`
2. `xpra attach ssh://$server/1 --clipboard=no -z0 --encoding=rgb`
3. Log in to a Windows (7, I believe) guest OS using vmware viewer.
4. Press Control C

**System Information (please complete the following information):**
 - Server OS: Debian 12
 - Client OS: Debian trixie
 - Xpra Server Version 5.0.6
 - Xpra Client Version 5.0.6

**Additional context**
```
2024-02-28 19:36:03,588 do_get_keycode (54, c)=54 (native keymap)
2024-02-28 19:36:03,588 process_key_action(('key-action', 5, 'c', True, ('control',), 99, '\x03', 54, 0)) server keycode=54, group=0
2024-02-28 19:36:03,588 set_keyboard_layout_group(0) config=KeyboardConfig(us / None / None), current keyboard group=0
2024-02-28 19:36:03,588 setting XKB layout group 0
2024-02-28 19:36:03,589 filtered_modifiers_set([])=set()
2024-02-28 19:36:03,589 filtered_modifiers_set(['control'])={'control'}
2024-02-28 19:36:03,589 make_keymask_match(['control']) current mask: set(), wanted: {'control'}, ignoring=54/['c'], keys_pressed={125: 'Alt_L'}
2024-02-28 19:36:03,589 change_mask(set(), False, remove) failed=[]
2024-02-28 19:36:03,589 keynames(control)=['Control_L', 'Control_R'], keycodes=[37, 109], nuisance=False, nuisance keys={'lock', 'mod2'}
2024-02-28 19:36:03,589 add control with keycode 37 did not work
2024-02-28 19:36:03,589  trying to unpress it!
2024-02-28 19:36:03,589 change_mask(add) ['control'] modifier 'control' using keycode 37, success: False
2024-02-28 19:36:03,589 add control with keycode 109 did not work
2024-02-28 19:36:03,589  trying to unpress it!
2024-02-28 19:36:03,589 change_mask(add) ['control'] modifier 'control' using keycode 109, success: False
2024-02-28 19:36:03,589 change_mask({'control'}, True, add) failed=['control']
2024-02-28 19:36:03,589 is_modifier(54) not found
2024-02-28 19:36:03,589 handle_key((5, True, 'c', 99, 54, ['control'], False, True))
2024-02-28 19:36:03,589 handle keycode pressing    54: key 'c'
2024-02-28 19:36:03,589 fake_key(54, True)
```

I'm confused by the Alt (I can re-check this if you're also suspicious that I copied the wrong block of the log).  Also, the keyboard layout changes ~5 times whenever the first character is pressed into the client window (only the last one is actually acted on, per the timeout in `_keys_changed`).  Additionally, (and I don't know if this is valuable or not), `xmodmap` doesn't show any bindings (!?!) when the affected window is focused. (i.e., `sleep 3 ; xmodmap` in a terminal, then send a keystroke to the vmware-horizon client window), then wait 3 seconds for the resulting xmodmap output.

`xdotool "Ctrl+c"` works in this setting, so while vmware is definitely doing something weird, it's not totally broken.

I can definitely test things and look into this further, I'm just at a loss where to go in the debugging step (i.e., is the modmap really supposed to be blank? should I be dumping (all?) the purported keyboard layout changes? what would I even be looking for? xdotool seems to get the mapping correct, what is it doing differently?)

**Comment by totaam** [comment]
```
keynames(control)=['Control_L', 'Control_R'], keycodes=[37, 109], nuisance=False, nuisance keys={'lock', 'mod2'}
add control with keycode 37 did not work
```
The keymap that xpra is working with has `Control_L = 37`, `Control_R = 109` but somehow this is not correct?
Perhaps vmware modified the keymap?

> `xmodmap` doesn't show any bindings

Yes, so something is definitely messed up. It should match the values above.
I'm going to close this as not-an-xpra-bug.

**Comment by aerusso** [comment]
Yeah, lobotomizing the `change_mask` logic by unconditionally accepting the keycode (by setting `modkeycode = keycode`  and breaking out of the loop, instead of un-pressing it) does get this working. (And thank you, your response was critical!)

If I can figure out a reasonable way to express the particular pathological behavior of this program and handle it gracefully, I'll submit a PR.  But I too am pessimistic.

**Comment by totaam** [comment]
@aerusso how about the commit above?
Run with `XPRA_VERIFY_MODIFIERS=0 xpra ...`

**Comment by aerusso** [comment]
Thanks! I'm trying to get at a more general solution (but I'm failing).  What I was hoping to do was basically this, but contingent on the modmap being empty (rather than checking a switch).  After looking at the code, though, it appears that it should already be doing (something very close to) that!  I.e., it should bail out of doing any of this careful mask-adjusting logic if the modmap is empty.

So, I investigated further:  in `_keys_changed` in `xpra/server/mixins/input.py`, the call to `ss.keys_changed()` is guarded by a `keymap_changing_timer`.  On my system, that is always an integer.  And it's only ever zero on the first call to it.  (Maybe there's an expectation it's one of the other timer implementations?)  So, I think that may be a bug.  It consequently is never calling the keymap-changing code (i.e., `compute_modifier_map`) . Bingo, I though!

Notwithstanding that, I went ahead and removed the `if not self.keymap_changing_timer` (since I only ever saw that function being called once per layout change, anyway).  I also added in a `sleep(0.02)` to sweep race conditions under the rug for debugging.  The keymap is empty (confirmed via `xmodmap`), but `grok_modifier_map` was showing up with all the modifier keys in it. So, either there's a race, and I need a longer `sleep()` there, or GDK is maybe lying to us about the keymap?

With this, I'm unfortunately out of time to look into this right now.  And I'm not sure I'm willing to trace into GDK to figure out what exactly is going on there.  I'll look into this more later, but it will unfortunately have to wait at least a ~week.

**Comment by totaam** [comment]
> since I only ever saw that function being called once per layout change, anyway

Lucky you!
(this function can fire many times on some platforms..)

---

The `keymap_changing_timer` value is used as a guard to prevent running the same `do_keys_changed` function multiple times in quick succession:
https://github.com/Xpra-org/xpra/blob/72c2632d5b9529020b462019a2e4f0487696aa10/xpra/server/gtk_server.py#L73-L75
Its value should be set when `keys_changed` is called for the first time in a while:
https://github.com/Xpra-org/xpra/blob/72c2632d5b9529020b462019a2e4f0487696aa10/xpra/server/gtk_server.py#L77
And it will be reset to zero on exit, allowing more `keys_changed` events to trigger it again.

The only slight issue that I can see is that we may miss a change event if it somehow fires whilst we're processing `do_keys_changed`.
This should not be possible as the events and the timers run in the same thread... but whatever, I've changed the order in 375c8f476f55cd78f6327115b05514049ac37794

**Comment by aerusso** [comment]
> @aerusso how about the commit above? Run with `XPRA_VERIFY_MODIFIERS=0 xpra ...`

I just tested this.  Indeed, these two commits do fix the issue  (I had to backporting 374175e7e7f546a0e63f5d62c42bd27609080bff to 5.0).

(I still want to see if I can figure out a more general solution, but it will obviously take some time.)
